### ISSUE TITLE
Noticeable performance downgrade from Python 3.10 to onwards versions

### ISSUE THREAD

**Issue by gi0baro** [PyO3/pyo3#4716]
Hi 👋 I'm not sure wether this makes sense as an issue or should be a discussion instead: to project maintainers, feel free to move this to a discussion if you conclude that's better.

Context: in the Granian project (an HTTP server) we recently introduced some [e2e benchmarks using different Python versions](https://github.com/emmett-framework/granian/blob/master/benchmarks/pyver.md), which show a ~30% performance degradation for some tests when comparing Python 3.10 to all other versions onwards (PyO3 0.22, cfg `pyo3_disable_reference_pool`).

Now, the specific tests showing this degradation involves some relatively simple code:
- we create both a `pyclass` and a `PyDict` objects (https://github.com/emmett-framework/granian/blob/c94e73e32a4865a011a4b659ef04bbc0a96e6fd4/src/wsgi/callbacks.rs#L24-L106)
- call a python function with those arguments (https://github.com/emmett-framework/granian/blob/c94e73e32a4865a011a4b659ef04bbc0a96e6fd4/granian/wsgi.py#L48-L58)
- that will call back a method of the first `pyclass` object (https://github.com/emmett-framework/granian/blob/c94e73e32a4865a011a4b659ef04bbc0a96e6fd4/src/wsgi/io.rs#L49-L57)

While I understand an e2e benchmark might suffer from a lot of additional noise when compared to a smaller _unit-benchmark_, and there's a lot more to consider (network stack in CPython stdlib, for example), I also believe, given other protocols involving asyncio and a bunch more stuff suffer from a very smaller degradation compared to the one I referenced, there might be something going on in PyO3 <-> CPython interop. Thus I have two main questions:
- is there any well-known difference from Python 3.11 and onwards in how PyO3 interact with the Python interpreter that might explain this?
- do you have any suggestions on how to investigate this in a more fine-grained way to help enlighten any other differences between Python versions that might play a role in this?

Thanks in advance 🙏

**Comment by davidhewitt** [comment]
Thanks for the questions. I don't have an immediate answer for you; to check I understand, Python 3.11 and up got ~30% slower?

Have you tried generating flame graphs (e.g. with [`samply`](https://github.com/mstange/samply)) to see if that gives a hint where the differences come from?

**Comment by gi0baro** [comment]
> to check I understand, Python 3.11 and up got ~30% slower?

@davidhewitt correct, here is a more direct comparison extracted from that bench:

| Python version | Total requests | RPS | avg latency | max latency |
| --- | --- | --- | --- | --- |
| 3.10 | 559148 | 55978 | 2.28ms | 24.538ms |
| 3.11 | 381549 | 38197 | 3.339ms | 24.674ms |
| 3.12 | 356792 | 35798 | 7.121ms | 64.292ms |
| 3.13 | 371324 | 37194 | 3.429ms | 18.313ms |

> Have you tried generating flame graphs (e.g. with [`samply`](https://github.com/mstange/samply)) to see if that gives a hint where the differences come from?

that makes sense. Let me plan some tests using sampling, I'll post here the findings.

**Comment by gi0baro** [comment]
@davidhewitt I tried using samply as you suggested to build flame graphs, but even with the following profile used in build

```toml
[profile.profiling]
inherits = "release"
debug = true
```

the stacks in the report just shows items as *0x12989b _granian.cpython-310-x86_64-linux-gnu.so*, so it's quite hard to spot on differences between 3.10 and 3.11 builds. Do you have any further suggestions on how to get full stacks on a PyO3 `cdylib` built library?

**Comment by gi0baro** [comment]
Btw, I re-run tests with PyO3 0.23 and they shows the same issue (ignore the absolute numbers as it's different hardware):

| Python version | RPS | avg latency | max latency |
| --- | --- | --- | --- |
| 3.9 | 132489 | 0.483ms | 1.653ms |
| 3.10 | 132521 | 0.482ms | 1.565ms |
| 3.11 | 64298 | 0.994ms | 1.816ms |
| 3.12 | 62252 | 2.054ms | 7.601ms |
| 3.13 | 63075 | 1.014ms | 1.692ms |

Might this be related to threads? The involved code has the main Python thread waiting on a `threading.Event` object, with 1 tokio thread dealing with I/O and sending/receiving stuff from a 2nd thread which interacts with Python code through a loop of `Python::with_gil(|py| { ... })` calls. I'm starting wondering there might be some difference in GIL acquisition from different threads after 3.10.

**Comment by davidhewitt** [comment]
Sorry for the long delay here. I was extremely busy for a few months and only now catching up with various things.

> the stacks in the report just shows items as 0x12989b _granian.cpython-310-x86_64-linux-gnu.so, so it's quite hard to spot on differences between 3.10 and 3.11 builds. Do you have any further suggestions on how to get full stacks on a PyO3 cdylib built library?

This... sounds like you're doing the right thing. I think maybe check you have also set `strip = false` in the profile?

**Comment by gi0baro** [comment]
> Sorry for the long delay here.

No worries :) 

> This... sounds like you're doing the right thing. I think maybe check you have also set `strip = false` in the profile?

Re-built with the following profile:

```toml
[profile.profiling]
inherits = "release"
debug = true
strip = false
```

but the stacks just show items as I previously reported (attaching just for reference) 🤷‍♂️

[prof311.a.json](https://github.com/user-attachments/files/19670617/prof311.a.json)
[prof310.a.json](https://github.com/user-attachments/files/19670616/prof310.a.json)

**Comment by davidhewitt** [comment]
Can you give me instructions how to run your benchmark? Maybe I can take a brief look (probably on Friday).

**Comment by gi0baro** [comment]
@davidhewitt here are the steps:

- clone the project @ https://github.com/emmett-framework/granian
- edit `Cargo.toml` and add the following profile:
    ```toml
    [profile.profiling]
    inherits = "release"
    debug = true
    strip = false
    ```
- build a virtualenv with the specific python versions (3.10, 3.11 in my example) and install `maturin` in each of them
- build the project in each venv using `maturin develop --profile profiling`
- launch the server with `samply record --save-only -o prof.json -- {venv_path}/bin/granian --interface wsgi --http 1 --no-ws --blocking-threads 1 benchmarks.app.wsgi:app`
- spawn a bunch of requests using `wrk` or similar. I personally use [rewrk](https://github.com/gi0baro/rewrk) with the following cmd: `rewrk -d 10s -c 128 --host http://localhost:8000/b10`

Mind that the performance downgrade is noticeable primarly on Linux; on MacOS the difference is quite subtle.
Also, thank you for your support! ❤️

**Comment by gi0baro** [comment]
> So a bit strange, this runs totally fine and has symbols for me. I'm not sure what's going on for you in the profiling (I'm using linux under WSL).

🤔 weird. Need to review my Linux env then.

> Good news is that with the symbols I was able to work out what's going on. `Python::with_gil` uses an API that will automatically GC the Python thread state. I thought this behaviour has existed longer, I forget exactly, but either way it looks like on every task on 3.11 you were setting up (and tearing down) a Python thread state.

Oh, I see. My ignorance in regards of the CPython internals led me to believe looping a `Python::with_gil()` call within the same thread would've kept the Python thread state as the id would be the same. Very nice to know.

> Applying this patch completely solves the perf regression: [...]

Thank you ❤ Checking the code change, so far it seems to solve the issue!
On a side note: I see you also changed the free-threaded version of the code to also call `py.allow_threads`. Is this actually necessary? I thought without the GIL, there's no need to release anything in threads?

> I'm not sure if we should document this behaviour, as the [underlying API](https://docs.python.org/3/c-api/init.html#c.PyGILState_Ensure) doesn't. However we might want to switch away from that API in PyO3 eventually (e.g. to support subinterpreters). So I'm not sure what work is suitable to do in PyO3 to improve this. 🤔

Maybe I can try to work to add a small section in the guide under the performance chapter? WDYT?

**Comment by davidhewitt** [comment]
Sorry for the slow reply as usual.

> On a side note: I see you also changed the free-threaded version of the code to also call `py.allow_threads`. Is this actually necessary? I thought without the GIL, there's no need to release anything in threads?

The API is probably misnamed now, however detaching from the interpreter is still a good idea to allow GC to run. The notes in the guide on free-threading cover this point.

> Maybe I can try to work to add a small section in the guide under the performance chapter? WDYT?

I think we probably do need something in the guide, though I got stuck on what to put because I think CPython is considering changes in this area (regarding the reference-counted thread states; this might be pushed as a user / PyO3 responsibility in the future).
